lib/msal-core/src/telemetry/TelemetryManager.ts

```diff
@@ -0,0 +1,131 @@
+import TelemetryEvent from "./TelemetryEvent";
+import { CompletedEvents, EventCount, EventCountByCorrelationId, InProgressEvents, TelemetryConfig, TelemetryPlatform } from "./TelemetryTypes";
+import DefaultEvent from "./DefaultEvent";
+import { platform } from "os";
+
+// for use in cache events
+const MSAL_CACHE_EVENT_VALUE_PREFIX = "msal.token";
+const MSAL_CACHE_EVENT_NAME = "msal.cache_event";
+
+const createEventKey = (event: TelemetryEvent): string => (
+    `${event.telemetryCorrelationId}-${event.eventId}-${event.eventName}`
+);
+
+export default class TelemetryManager {
+
+    // correlation Id to list of events
+    private completedEvents: CompletedEvents = {};
+    // event key to event
+    private inProgressEvents: InProgressEvents = {};
+    // correlation id to map of eventname to count
+    private eventCountByCorrelationId: EventCountByCorrelationId = {};
+
+    //Implement after API EVENT
+    private onlySendFailureTelemetry: boolean = false;
+    private telemetryPlatform: TelemetryPlatform;
+    private clientId: string;
+    private telemetryCallback: Function = null;
+
+    constructor(config: TelemetryConfig, cb: Function) {
+        // TODO THROW if bad options
+        this.telemetryPlatform = config.platform;
+        this.clientId = config.clientId;
+        this.onlySendFailureTelemetry = config.onlySendFailureTelemetry;
+        // TODO, when i get to wiring this through, think about what it means if
+        // a developer does not implement telem at all, we still instrument, but cb can be
+        // optional?
+        this.telemetryCallback = cb;
```

Comment:
You could default it to a noop function, which would remove the need downstream to always be checking the truthiness of the `telemetryCallback` function.

Reply:
I suppose unless you don't want the code after those checks to run at all if the function isn't passed.

Reply:
Yeah I think for now that if the callback is not passed, we don't want to track any of this data.